### ISSUE TITLE
[flang] Error not detected for statement function dummy name with EXTERNAL attribute used in data context

### ISSUE THREAD

**Issue by ejose02** [llvm/llvm-project#203820]
Flang accepts a program unit where a statement function’s dummy argument is declared with the EXTERNAL attribute, but the same identifier appears in the statement-function definition in a data context. That is an inconsistent classification: the dummy cannot simultaneously be treated as an external procedure in the header and as a numeric operand in the expression. LLVM Flang currently compiles the example without any diagnostic and produces an executable.

Expected output: Compile time error with proper error message

Actual output: No error and no warning; compilation succeeds.

Reproducer:

```Fortran
real function fun(x)
  real, external :: z
  sf(z) = z + 10
  fun = sf(x)
end function fun
print *, fun(30.0)
end
```

**Comment by kwyatt-ext** [comment]
I don't believe this is a bug — this is required behavior.

A statement function's dummy arguments have a scope limited to the statement function statement itself. Per F2018 Section 15.6.4.3, each statement function dummy argument has "the same type and type parameters" as an entity of that name would have in the host scoping unit.

EXTERNAL is not a type parameter. Type parameters are only the KIND and LEN parameters that parameterize a type (F2018 Section 7.2.2), e.g. the 4 in REAL(4) or the 20 in CHARACTER(20). EXTERNAL is an attribute (8.5.9), so it is deliberately not inherited.

In this example the host z is REAL, EXTERNAL.  Its type/type parameters (REAL(4)) flow into the statement function dummy.  Its EXTERNAL attribute does not.  The the two zs are distinct entities — the host z is a ProcEntity with EXTERNAL, while the statement function dummy z is a plain REAL(4) ObjectEntity.

This matches the standard, so there's nothing to fix.  @ejose02 Please close this issue if you agree.

**Comment by eugeneepshteyn** [comment]
On the other hand, NAG, nvfortran, gfortran, and ifx reject this program. 

@kwyatt-ext , what about F2023, 19.4, paragraph 2: "The name of a statement entity shall not be the same as an accessible global identifier or local identifier of class 10 (1) (19.3.1), except for a common block name or a scalar variable name." Wouldn't `z` be a statement entity here? And being marked as external procedure, wouldn't `z` be a global identifier?

**Comment by kwyatt-ext** [comment]
@eugeneepshteyn is correct.  I missed that scope requirement.

@ejose02 Do you intend to fix this?  If not, I'll grab it since I already dug into it some.

**Comment by ejose02** [comment]
> @eugeneepshteyn is correct.  I missed that scope requirement.
> 
> @ejose02 Do you intend to fix this?  If not, I'll grab it since I already dug into it some.

You can go ahead with the fix.
